src/ast/diagnostic.cpp

```diff
@@ -18,16 +20,37 @@ void Diagnostics::emit(std::ostream& out, Severity s) const
 
 void Diagnostics::emit(std::ostream& out, Severity s, const Diagnostic& d) const
 {
-  const auto& loc = d.loc();
+  // Build our set of messages.
+  std::vector<std::pair<std::string, SingleLocation>> msgs;
+  for (auto loc = d.loc(); loc; loc = loc->parent->loc) {
+    if (loc->parent.has_value()) {
+      msgs.emplace_back(loc->parent->msg.str(), loc->current);
+    } else {
+      msgs.emplace_back(d.msg(), loc->current);
+      break;
+    }
+  }
+  std::ranges::reverse(msgs);
+
   switch (s) {
     case Severity::Warning:
-      LOG(WARNING, loc.source_location(), loc.source_context(), out) << d.msg();
+      if (msgs.empty()) {
+        LOG(WARNING, out) << d.msg();
```

Comment:
JC, when does it happen that the diagnostic has an error without a location?

Reply:
There are tests that construct a location with Location(). This used to result in a non-sensical line/column which was ignored by the log. It’s actually better like this, since the the empty location is much more explicit (and maybe I should add asserts that the locations logged are valid, and drop the default SingleLocation constructor).